mpmath/functions/functions.py

```diff
@@ -164,11 +164,80 @@ def expm1(ctx, x):
+    # Note that all cases could by handled by log(1+c) provided the
+    # add is done exactly. Our aim here is be much faster than that,
+    # especially when |c| is small.
+    c = ctx.convert(x)
+    cmag = ctx.mag(c)
+    a, b = c.real, c.imag
+    wp = ctx.prec
+    if cmag >= -wp:
+        # |c| isn't very small. We call log(1+c) instead, but
+        # are careful about the precision used by the add. The
+        # real part of the result is log(|c+1|). That's
+        # determined by 1 + 2*a + a**2 + b**2, and the add has
+        # to preserve enough info so that no important bits of
+        # that sum are lost. It doesn't matter to this that 2*a,
+        # a**2, etc, are not computed explicitly here: we're
+        # deducing how many bits have to be present in the sum
+        # for log() to "reverse engineer" the value of 2*a +
+        # a**2 + b**2 to `prec` good bits,
+        if cmag < 4:
+            # |c| isn't very small, or large.
+            if ctx.mag(a) > ctx.mag(b):
+                # `a` already contributes the most to c's norm.
+                # After adding 1, it will utterly dominate it.
+                # We only need enough extra precision to avoid
+                # losing any of a's `prec` most significant bits
+                # when addiog, `b**2` is too small to matter.
+                wp *= 2
+            else:
+                # b**2 is the larger of the square terms. The
+                # smallest b can be is about 2**-prec, so the
+                # smallest b**2 can be is about 2**(-2*prec). So
+                # for a bit to matter compared to b**2, it has
+                # to be at least about 2**(-3*prec). Bits of 2*a
+                # (if any) >= 2**(-3*prec) will be preserved if
+                # we use 3*prec bits for the add.
+                wp *= 3
+        # Else (cmag >= 4), |c+1| >= |c| - 1 is so large that
+        # working precision is fine (although that takes some
+        # careful analysis for cmag=4, given that .mag() _may_
+        # return a rexult too large by 2), So leave wp alone.
+        arg = ctx.fadd(1.0, c, prec=wp)
+        result = ctx.log(arg)
+    else:
+        # Else c is "very small", and we use a series expansion,
+        # c - c**2/2. The real part of that is a+(b*b-a*a)/2,
+        # and the imag part b-a*b. Given that cmag < -prec, it
+        # can be shown that "a*b" is numerically insignifcant in
+        # the imag part, and _usually_ the "a*a/2" in the real
+        # part. What remains is cheap to compute. In the real
+        # part, though, if `a` is negative, the remaining
+        # a+b**2/2 can suffer massive cancellation - even total.
+        real = a + b*b*0.5 # usually the real part of the result
+        if (a < 0.0
+            and ctx.mag(real) <= ctx.mag(a) - LOG1P_EXTRAPREC):
+            # The guard bits were lost to cancellation. Rare. At
+            # the contrived
+            # -1.999999873062092e-40+1.999999936531045e-20j
```

Comment:
Should add this input to the new tests - it's, by far, "the hardest" input I found.